In [313]:
import re
import pandas as pd
sys.path.append('../')
import src.limpieza as lmp
import numpy as np

In [314]:
pd.options.display.max_columns = None

In [315]:
revius = open("output1.html", "r", encoding='utf-8').readlines()

In [316]:
civitatis = [' '.join(str(i) for i in revius)]

In [317]:
print(len(civitatis))

1


In [318]:
corte = '\ue9ce\ue9ce\ue9ce\ue9ce\ue9ce\n'
civitatis_ = civitatis[0].split(corte)

In [319]:
civitatis_revius = [i for i in civitatis_[2::2]]

In [320]:
civitatis_revius_limpio = [i.split('\n') for i in civitatis_revius]

In [321]:
longitud = [len(i) for i in civitatis_revius_limpio]
long_ = set(longitud)
print(long_)

{4, 5, 6, 7, 8, 9, 10, 11, 12, 13}


In [322]:
for i in range(0,14):
    for a in civitatis_revius_limpio:
        if len(a)==i:
            print(a)
            break

[' 11 / Mar / 2022', ' A', ' Anónimo', ' ']
[' 10 / Mar / 2022', ' D', ' Diana Cruz Torres', ' Toa Alta, Puerto Rico', ' ']
[' 11 / Mar / 2022', ' A', ' Ainhoa', ' España', ' Viajó en pareja', ' ']
[' 11 / Mar / 2022', ' A', ' Ana', ' Madrid, España', ' Victor es encantador, te explica todo al detalle y te resuelve todas las preguntas, volvería a repetir, me a encantado.', ' Viajó con hijos mayores', ' ']
[' 11 / Mar / 2022', ' G', ' Graciela Arguelles', ' Guanajuato, Montenegro', ' Excelente recorrido ️️️️️', ' Opinión traducida. Mostrar en idioma original', ' Viajó con amigos', ' ']
[' 08 / Mar / 2022', ' A', ' Anónimo', ' 16195 opiniones', ' Mostrando 37-48', ' primera anterior', ' Página 2 Página 3 Página 4 Página 5 Página 6', ' siguiente última', ' ']
[' 10 / Feb / 2022', ' A', ' Anónimo', ' Viajó solo', ' 16195 opiniones', ' Mostrando 469-480', ' primera anterior', ' Página 38 Página 39 Página 40 Página 41 Página 42', ' siguiente última', ' ']
[' 11 / Mar / 2022', ' M', ' María J

In [323]:
civitatis_limpio_oi = lmp.pulir(civitatis_revius_limpio)
civitatis_reluciente = lmp.quitar_respuestas(civitatis_limpio_oi)


In [324]:
longitud = [len(i) for i in civitatis_reluciente]
long_ = set(longitud)
print(long_)

{2, 3, 4, 5}


In [325]:
len(civitatis_reluciente)

16231

In [326]:
dict_revius = lmp.dict_civitatis_completo(civitatis_reluciente)

In [327]:
len(dict_revius['fecha']) == len(dict_revius['nombre']) == len(dict_revius['procedencia']) == len(dict_revius['opinion']) == len(dict_revius['viajo_con'])

True

In [328]:
df_civitatis = pd.DataFrame(dict_revius)

In [329]:
df_civitatis.head(5)

fecha                     nombre      procedencia  \
0   11 / Mar / 2022                        Ana   Madrid, España   
1   11 / Mar / 2022                 Montserrat           España   
2   11 / Mar / 2022                     Ainhoa           España   
3   11 / Mar / 2022   Patricia Suárez Carrillo   Madrid, España   
4   11 / Mar / 2022                    Anónimo                    

                                             opinion                 viajo_con  
0   Victor es encantador, te explica todo al deta...   Viajó con hijos mayores  
1   Casi 3 horas de tour pero Víctor, el guía que...                Viajó solo  
2                                                              Viajó en pareja  
3   A mi pareja y a mi nos ha encantado! Nuestro ...           Viajó en pareja  
4

In [330]:
df_civitatis.shape

(16231, 5)

In [331]:
df_civitatis.duplicated().sum()

4690

We can se that we have quite a lot of duplicates, actually around 29% of the rows seem to be replicated but, is this completely truth?

In [332]:
df_civitatis[df_civitatis.duplicated() == True]

fecha                     nombre      procedencia  \
12      11 / Mar / 2022                        Ana   Madrid, España   
13      11 / Mar / 2022                 Montserrat           España   
14      11 / Mar / 2022                     Ainhoa           España   
15      11 / Mar / 2022   Patricia Suárez Carrillo   Madrid, España   
16      11 / Mar / 2022                    Anónimo                    
...                 ...                        ...              ...   
16177   04 / May / 2019                    Anónimo           España   
16190   06 / Mar / 2019                    Anónimo           España   
16196   15 / Feb / 2019                    Anónimo           España   
16198   08 / Feb / 2019                    Anónimo           España   
16221   14 / Oct / 2018                    Anónimo           España   

                                                 opinion  \
12      Victor es encantador, te explica todo al deta...   
13      Casi 3 horas de tour pero Víctor, el guía que...   
14                                                         
15      A mi pareja y a mi nos ha encantado! Nuestro ...   
16                                                         
...                                                  ...   
16177                                                      
16190                                                      
16196                                                      
16198                                                      
16221                                                      

                      viajo_con  
12      Viajó con hijos mayores  
13                   Viajó solo  
14              Viajó en pareja  
15              Viajó en pareja  
16                               
...                         ...  
16177                            
16190                            
16196                            
16198                            
16221                            

[4690 rows x 5 columns]

In [333]:
df_civitatis.iloc[13,3]

' Casi 3 horas de tour pero Víctor, el guía que nos ha acompañado ha sido un encanto. Hemos andado muchísimo pero ha merecido la pena porque se nota que le gusta mucho su trabajo y se ha hecho muy ameno y emocionante. Sin duda un tour 100% recomendado con este guía. Gracias por hacernos pasar una mañana increíble.'

In [334]:
df_civitatis[df_civitatis['opinion'] == 'Casi 3 horas de tour pero Víctor, el guía que nos ha acompañado ha sido un encanto. Hemos andado muchísimo pero ha merecido la pena porque se nota que le gusta mucho su trabajo y se ha hecho muy ameno y emocionante. Sin duda un tour 100% recomendado con este guía. Gracias por hacernos pasar una mañana increíble.']


Empty DataFrame
Columns: [fecha, nombre, procedencia, opinion, viajo_con]
Index: []

We can be absolutely sure that the we have doubled values only if we check the column 'opinion', as it's the only one unrepeatable. On the other hand, we can easely have several people from Spain doing a tour the same day who wouldn't leave their names, so all of them will appear as 'Anonimo' and 'España' with the other columns in blank, and pandas will take them as duplicates while they actually were different clients.
So let's analize this in two different data frames. First we'll analize the revius with the 'opinion' column filled.

In [335]:
dict_revius_5 = lmp.dict_civitatis_5(civitatis_reluciente)

In [336]:
df_civitatis_5 = pd.DataFrame(dict_revius_5)

In [337]:
df_civitatis_5.shape

(8953, 5)

In [338]:
df_civitatis_5.head(5)

fecha                     nombre        procedencia  \
0   11 / Mar / 2022                        Ana     Madrid, España   
1   11 / Mar / 2022                 Montserrat             España   
2   11 / Mar / 2022   Patricia Suárez Carrillo     Madrid, España   
3   11 / Mar / 2022                      Saray   Alicante, España   
4   11 / Mar / 2022       Tania Carrera Garcia             España   

                                             opinion                 viajo_con  
0   Victor es encantador, te explica todo al deta...   Viajó con hijos mayores  
1   Casi 3 horas de tour pero Víctor, el guía que...                Viajó solo  
2   A mi pareja y a mi nos ha encantado! Nuestro ...           Viajó en pareja  
3                   El guia del tour, Ponce, Genial.          Viajó con amigos  
4   Mónica ha sido una guía genial, nos ha explic...          Viajó con amigos

In [339]:
df_civitatis_5.duplicated().sum()

2114

In [340]:
df_civitatis_5[df_civitatis_5.duplicated() == True]

fecha                     nombre               procedencia  \
9      11 / Mar / 2022                        Ana            Madrid, España   
10     11 / Mar / 2022                 Montserrat                    España   
11     11 / Mar / 2022   Patricia Suárez Carrillo            Madrid, España   
12     11 / Mar / 2022                      Saray          Alicante, España   
13     11 / Mar / 2022       Tania Carrera Garcia                    España   
...                ...                        ...                       ...   
8836   15 / Jun / 2017                      Pablo   Buenos Aires, Argentina   
8837   15 / Jun / 2017            Sofia Hernandez      Guatemala, Guatemala   
8838   15 / Jun / 2017                       Luis            Madrid, España   
8839   11 / Jun / 2017                      Maria            Madrid, España   
8866   25 / Sep / 2021              Rita Cantante           Porto, Portugal   

                                                opinion  \
9      Victor es encantador, te explica todo al deta...   
10     Casi 3 horas de tour pero Víctor, el guía que...   
11     A mi pareja y a mi nos ha encantado! Nuestro ...   
12                     El guia del tour, Ponce, Genial.   
13     Mónica ha sido una guía genial, nos ha explic...   
...                                                 ...   
8836    Excelente paseo y pablo nuestro guia fantastico   
8837   Me gusto mucho, amable y dispuesto a contesta...   
8838   Una guía agradable de unas dos horas en las q...   
8839   Nos ha encantado, muy amena y Pablo el guía s...   
8866   Muy interactivo, muy divertido. ¡Alex sabía m...   

                     viajo_con  
9      Viajó con hijos mayores  
10                  Viajó solo  
11             Viajó en pareja  
12            Viajó con amigos  
13            Viajó con amigos  
...                        ...  
8836           Viajó en pareja  
8837           Viajó en pareja  
8838   Viajó con hijos mayores  
8839           Viajó en pareja  
8866           Viajó en pareja  

[2114 rows x 5 columns]

In [341]:
df_civitatis_5.iloc[0, 3]

' Victor es encantador, te explica todo al detalle y te resuelve todas las preguntas, volvería a repetir, me a encantado.'

In [342]:
df_civitatis_5[df_civitatis_5['opinion'] == ' Victor es encantador, te explica todo al detalle y te resuelve todas las preguntas, volvería a repetir, me a encantado.']

fecha nombre      procedencia  \
0    11 / Mar / 2022    Ana   Madrid, España   
9    11 / Mar / 2022    Ana   Madrid, España   
26   11 / Mar / 2022    Ana   Madrid, España   

                                              opinion  \
0    Victor es encantador, te explica todo al deta...   
9    Victor es encantador, te explica todo al deta...   
26   Victor es encantador, te explica todo al deta...   

                   viajo_con  
0    Viajó con hijos mayores  
9    Viajó con hijos mayores  
26   Viajó con hijos mayores

And here we can see that that some reviews are repeated up to three times... 
Anyway, from the total number of rows, 8953, this means that 23,6% of them are duplicated, a value significantly lower than the 29% we got for whole data frame.
Now let's drop the duplicates in this one and then we will work with the reviews that don't have the 'opinion' column filled.

In [343]:
df_civitatis_5 = df_civitatis_5.drop_duplicates()
df_civitatis_5.duplicated().sum()

0

In [344]:
df_civitatis_5[df_civitatis_5['opinion'] == ' Victor es encantador, te explica todo al detalle y te resuelve todas las preguntas, volvería a repetir, me a encantado.']

fecha nombre      procedencia  \
0   11 / Mar / 2022    Ana   Madrid, España   

                                             opinion                 viajo_con  
0   Victor es encantador, te explica todo al deta...   Viajó con hijos mayores

All the reviews whithout 'opinion':

In [345]:
dict_civitatis_2_4 = lmp.dict_civitatis_2_4(civitatis_reluciente)

In [346]:
df_civitatis_2_4 = pd.DataFrame(dict_civitatis_2_4)

In [347]:
df_civitatis_2_4.shape

(7278, 5)

In [348]:
df_civitatis_2_4.head(5)

fecha                      nombre      procedencia opinion  \
0   11 / Mar / 2022                      Ainhoa           España           
1   11 / Mar / 2022                     Anónimo                            
2   11 / Mar / 2022   María José Campillo Mateo   Murcia, España           
3   11 / Mar / 2022                      Ainhoa           España           
4   11 / Mar / 2022                     Anónimo                            

          viajo_con  
0   Viajó en pareja  
1                    
2   Viajó en pareja  
3   Viajó en pareja  
4

In [349]:
df_civitatis_2_4.duplicated().sum()

2576

If we take the reviews without 'opinion', we can see that the number of duplicates is signifincantly higher, around a 35%. This is due to the fact commented above that there can easely be two different clients with the same name (or name and origin) the same day.
Taking this into account, we won't drop all the duplicates, as this will alter significantly the data and we will be earasing real tourist from our data frame. But as we cannot state for sure which rows are duplicates and which ones are not, we will ramdomly erase 23.6% of them, as this seems to be the real percentage of duplicates that the algorithm generates; this means that we are going to drop 1717 rows in stead of the 2576 that pandas duplicated() method proposes.

In [350]:
df_civitatis_2_4 = lmp.drop_tourist(df_civitatis_2_4)

And we check that 23.6% of the rows are gone

In [351]:
df_civitatis_2_4.shape

(5561, 5)

Now that we have both datesets cleaned of duplicates, we can concatenate them

In [352]:
pd_civitatis_clean= pd.concat([df_civitatis_5, df_civitatis_2_4])

In [353]:
pd_civitatis_clean.shape

(12400, 5)

Above we verify that 23.6% of the rows are gone. From 16231 at the begining, now we have 12400.

Time to clean the data

In [354]:
pd_civitatis_clean.sample(5)

fecha             nombre      procedencia  \
2242   28 / Jun / 2021            Anónimo           España   
6659   19 / Abr / 2019          Alexander   Ferrol, España   
5424   20 / Mar / 2019            Anónimo           España   
7475   18 / Nov / 2018      Antonia Jesus   Madrid, España   
5808   17 / Ene / 2019   Rodrigo Lagrasta    Elche, España   

                                                opinion         viajo_con  
2242                                                                       
6659         Excelente la pasamos muy bien Pablo GENIAL   Viajó en pareja  
5424                                                                       
7475   Pablo un chico súper profesional, muy simpáti...        Viajó solo  
5808                                                           Viajó solo

First, we will take away the spaces at the begining and end of the values

In [355]:
df_civitatis_clean = lmp.values_sin_espacios(df_civitatis)

Now we will deal with the dates ('fecha' column), so they will be in date-time format, as this will be the main criteria we will use to organize the tourist

In [356]:
df_civitatis_clean.iloc[0,0]

'11 / Mar / 2022'

In [357]:
df_civitatis_clean = lmp.date_time(df_civitatis)

In [358]:
df_civitatis_clean.iloc[0,0]

Timestamp('2022-03-11 00:00:00')

In [359]:
df_civitatis_clean.sort_values(by=['fecha']).head(5)

fecha           nombre           procedencia  \
15955 2017-06-09            Charo     Salamancs, España   
15947 2017-06-09            Charo     Salamancs, España   
15954 2017-06-11            Maria        Madrid, España   
15946 2017-06-11            Maria        Madrid, España   
15952 2017-06-15  Sofia Hernandez  Guatemala, Guatemala   

                                                 opinion        viajo_con  
15955                                                     Viajó en pareja  
15947                                                     Viajó en pareja  
15954  Nos ha encantado, muy amena y Pablo el guía su...  Viajó en pareja  
15946  Nos ha encantado, muy amena y Pablo el guía su...  Viajó en pareja  
15952  Me gusto mucho, amable y dispuesto a contestar...  Viajó en pareja

In [360]:
df_civitatis_clean.to_pickle(r'C:\Users\mituc\Ironhack\Curso\IronLabs\Proyecto-1\data\civitatis_clean.pkl')


In [361]:
df_turistas_fecha = df_civitatis_clean.set_index('fecha')

In [362]:
df_numero_turistas_fecha = df_turistas_fecha.groupby([(df_turistas_fecha.index.year),(df_turistas_fecha.index.month)]).count()

In [363]:
df_numero_turistas_fecha.tail(5)

nombre  procedencia  opinion  viajo_con
fecha fecha                                         
2021  11        553          553      553        553
      12        678          678      678        678
2022  1         386          386      386        386
      2         470          470      470        470
      3         203          203      203        203

In [364]:
df_numero_turistas_fecha = df_numero_turistas_fecha.drop(['procedencia', 'opinion', 'viajo_con'], axis=1).rename(columns={'nombre':'numero_turistas'})

In [375]:
df_numero_turistas_fecha.head(5)

numero_turistas
fecha fecha                 
2017  6                   19
      7                   26
      8                   82
      9                   58
      10                  34

In [376]:
df_numero_turistas_fecha.to_pickle(r'C:\Users\mituc\Ironhack\Curso\IronLabs\Proyecto-1\data\numero_turistas_fecha.pkl')
